# Filtrado por condiciones con `Pandas`.

In [2]:
import pandas as pd
import numpy as np

In [3]:
datos = pd.read_csv('../Datos/bestsellers.csv')
datos.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Filtremos las observaciones que cumplen que el año sea mayor a 2016.

In [5]:
datos['Year'] > 2016

0      False
1      False
2       True
3       True
4       True
       ...  
545     True
546    False
547     True
548     True
549     True
Name: Year, Length: 550, dtype: bool

Nótese que el retorno es una lista *booleana*. Para obtener la tabla, podemos pasarla como un *slicing* para el objeto donde tenemos los datos:

In [6]:
datos[datos['Year'] > 2016]

Name  \
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
7                       A Gentleman in Moscow: A Novel   
8        A Higher Loyalty: Truth, Lies, and Leadership   
..                                                 ...   
544                                             Wonder   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
7                 Amor Towles          4.7    19699     15  2017      Fiction  
8                 James Comey          4.7     5983      3  2018  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
544             R. J. Palacio          4.8    21625      9  2017      Fiction  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[150 rows x 7 columns]

Nótese que se devuelven 150 registros únicamente, que son todas aquellas observaciones que cumplen con la condición de que el año asociado es mayor estrictamente a 2016.

Filtremos ahora los libros que sean de ficción y que su fecha de publicación sea del 2019 en adelante.

In [11]:
datos[(datos['Genre'] == 'Fiction') & (datos['Year'] > 2018)]

Name                 Author  \
41            Brown Bear, Brown Bear, What Do You See?        Bill Martin Jr.   
84   Dog Man: Brawl of the Wild: From the Creator o...             Dav Pilkey   
85   Dog Man: Fetch-22: From the Creator of Captain...             Dav Pilkey   
86   Dog Man: For Whom the Ball Rolls: From the Cre...             Dav Pilkey   
127                               Giraffes Can't Dance          Giles Andreae   
145                                     Goodnight Moon    Margaret Wise Brown   
155  Harry Potter and the Goblet of Fire: The Illus...          J. K. Rowling   
179                       If Animals Kissed Good Night      Ann Whitford Paul   
252                          Oh, the Places You'll Go!              Dr. Seuss   
267             Player's Handbook (Dungeons & Dragons)       Wizards RPG Team   
303             Strange Planet (Strange Planet Series)         Nathan W. Pyle   
397                             The Guardians: A Novel           John Grisham   
462                                 The Silent Patient       Alex Michaelides   
482                        The Very Hungry Caterpillar             Eric Carle   
489                   The Wonderful Things You Will Be  Emily Winfield Martin   
491                                   The Wonky Donkey            Craig Smith   
509                              To Kill a Mockingbird             Harper Lee   
529  What Should Danny Do? (The Power to Choose Ser...              Adir Levy   
534                            Where the Crawdads Sing            Delia Owens   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)            Jeff Kinney   

     User Rating  Reviews  Price  Year    Genre  
41           4.9    14344      5  2019  Fiction  
84           4.9     7235      4  2019  Fiction  
85           4.9    12619      8  2019  Fiction  
86           4.9     9089      8  2019  Fiction  
127          4.8    14038      4  2019  Fiction  
145          4.8     8837      5  2019  Fiction  
155          4.9     7758     18  2019  Fiction  
179          4.8    16643      4  2019  Fiction  
252          4.9    21834      8  2019  Fiction  
267          4.8    16990     27  2019  Fiction  
303          4.9     9382      6  2019  Fiction  
397          4.5    13609     14  2019  Fiction  
462          4.5    27536     14  2019  Fiction  
482          4.9    19546      5  2019  Fiction  
489          4.9     8842     10  2019  Fiction  
491          4.8    30183      4  2019  Fiction  
509          4.8    26234      7  2019  Fiction  
529          4.8     8170     13  2019  Fiction  
534          4.8    87841     15  2019  Fiction  
545          4.9     9413      8  2019  Fiction